File: als1/linked_list/linked_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
        else:
            self.tail.next = item
        self.tail = item

    def print_all_nodes(self):
        node = self.head
        while node != None:
            print(node.value)
            node = node.next

    def find(self, val):
        node = self.head
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    def find_all(self, val):
        found_nodes = []
        node = self.head
        while node is not None:
            if node.value == val:
                found_nodes.append(node)
            node = node.next
        return found_nodes

    def delete(self, val, all=False):
        if self.head is None:
            return None

        while self.head.value == val:
            self.head = self.head.next
            if self.head is None:
                self.tail = None
                return
            if not all:
                return

        prev = self.head
        node = self.head
        while node is not None:
            if node.value == val:
                if all:
                    prev.next = (
                        node.next
                        if node.next is not None and node.next.value != val
                        else prev.next
                    )
                    node = node.next
                    if node is None:
                        prev.next = None
                        self.tail = prev
                    continue
                else:
                    prev.next = node.next
                if prev.next is None:
                    self.tail = prev
                if not all:
                    return

            prev = node
            node = node.next
        return None

    def clean(self):
        self.head = None
        self.tail = None

    def len(self):
        res = 0
        node = self.head
        while node is not None:
            res += 1
            node = node.next
        return res

    def insert(self, afterNode, newNode):
        if self.tail is None:
            self.tail = newNode
        if afterNode is None:
            buffer_node = self.head
            self.head = newNode
            self.head.next = buffer_node
            return
        node = self.head
        while node is not None:
            if node == afterNode:
                buffer_node = node.next
                node.next = newNode
                newNode.next = buffer_node
                if newNode.next is None:
                    self.tail = newNode
                return
            node = node.next

File: asd_go/twosidelinkedlist/linked_list.go
package main

import (
	"errors"
	"os"
	"reflect"
)

type Node struct {
	prev  *Node
	next  *Node
	value int
}

type LinkedList2 struct {
	head *Node
	tail *Node
}

func (l *LinkedList2) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
		l.head.next = nil
		l.head.prev = nil
	} else {
		l.tail.next = &item
		item.prev = l.tail
	}

	l.tail = &item
	l.tail.next = nil
}

func (l *LinkedList2) Count() int {
	count := 0
	node := l.head
	for node != nil {
		count += 1
		node = node.next
	}
	return count
}

// error не nil, если узел не найден
func (l *LinkedList2) Find(n int) (Node, error) {
	node := l.head
	for node != nil {
		if node.value == n {
			return *node, nil
		}
		node = node.next
	}
	return Node{}, errors.New("node not found")
}

func (l *LinkedList2) FindAll(n int) []Node {
	var nodes []Node
	node := l.head
	for node != nil {
		if node.value == n {
			nodes = append(nodes, *node)
		}
		node = node.next
	}
	return nodes
}

func (l *LinkedList2) Delete(n int, all bool) {
	current := l.head
	for current != nil {
		if current.value == n {
			if current.prev != nil {
				current.prev.next = current.next
			} else {
				l.head = current.next
			}
			if current.next != nil {
				current.next.prev = current.prev
			} else {
				l.tail = current.prev
			}
			if !all {
				return
			}
		}
		current = current.next
	}
}

func (l *LinkedList2) Insert(after *Node, add Node) {
	if after == nil {
		add.next = l.head
		add.prev = nil
		if l.head != nil {
			l.head.prev = &add
		}
		l.head = &add
		if l.tail == nil {
			l.tail = &add
		}
		return
	}

	add.prev = after
	add.next = after.next
	after.next = &add

	if add.next != nil {
		add.next.prev = &add
	} else {
		l.tail = &add
	}
}

func (l *LinkedList2) InsertFirst(first Node) {
	node := l.head
	if node == nil {
		l.head = &first
		l.tail = &first
		return
	}

	l.head = &first
	first.next = node
	node.prev = &first
}

func (l *LinkedList2) Clean() {
	l.head = nil
	l.tail = nil
}
